src/src/com/microsoft/aad/adal/DefaultTokenCacheStore.java

```diff
@@ -98,11 +98,14 @@ public DefaultTokenCacheStore(Context context) {
         if (mPrefs == null) {
             throw new IllegalStateException(ADALError.DEVICE_SHARED_PREF_IS_NOT_AVAILABLE.getDescription());
         }
-
+        
+        // Check upfront when initializing DefaultTokenCacheStore. 
+        // If it's under API 18 and secretkey is not provided, we should fail upfront to inform 
+        // notify developers. 
         try {
-            getStorageHelper().loadSecretKeyForAPI();
-        } catch (IOException | GeneralSecurityException e) {
-            throw new IllegalStateException("Failed to get private key from AndroidKeyStore", e);
+            getStorageHelper().loadSecretKeyForEncryption();
+        } catch (final IOException | GeneralSecurityException e) {
+            Logger.v(TAG, "Fail to create keys from android keystore.");
```

Comment:
only check if the key is supplied for API < 18, also extract it into a separate separate method. validatekeySetting